### ISSUE TITLE
`needless_range_loop` being overzealous

### ISSUE THREAD

**Issue by kaathewisegit** [rust-lang/rust-clippy#15596]
### Description

`needless_range_loop` heavily discourages the use of indexes, even when the body isn't solely about indexing.  I have the following code:

```rust
		for i in 0..self.num_leaves() {
			self.heights.set(i, heights[i]);
			self.heights.accept_element(i);
		}
```

Which clippy suggests to change it to `for (i, <item>) in heights.iter().enumerate().take(self.num_leaves())`, which would be be much less readable:

```rust
		for (i, height) in
			heights.iter().enumerate().take(self.num_leaves())
		{
			self.heights.set(i, *height);
			self.heights.accept_element(i);
		}
```

The iteration code becomes very heavy just to save one bounds check on indexing.  I feel that the code complexity limit for triggering this warning should be lower.  Or that there should be an opt-in more eager version

### Version

```text
rustc 1.89.0 (29483883e 2025-08-04)
binary: rustc
commit-hash: 29483883eed69d5fb4db01964cdf2af4d86e9cb2
commit-date: 2025-08-04
host: x86_64-unknown-linux-musl
release: 1.89.0
LLVM version: 20.1.7
```

**Comment by seandepagnier** [comment]
I first proposed to ignore this pedatic check instead of using basically unreadable iterators.  After they changed the code anyway and actually completely broke its function I didn't wish to maintain this garbage so  I rewrote the whole program in python using 40% fewer lines of code and it ran fast enough anyway, so they terminated me.

Your cases are not even half as bad as my cases.

Good luck trying to convince "cargo cult" members there is anything wrong with the clippy ritual.

**Comment by KisaragiEffective** [comment]
You could `#[expect(clippy::needless_range_loop)]` it...

**Comment by kaathewisegit** [comment]
Yeah, I've been using `expect` thus far.  I just think it'd be nice to have more fine-grained options.  One for very trivial cases and one for every place where it can save a bounds check
